ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/WhereInfo.java

```diff
@@ -185,38 +187,67 @@ public boolean isWindowed() {
     return getWindowBounds().isPresent();
   }
 
-  private static List<Object> extractKeysFromInPredicate(
+  private static List<GenericKey> extractKeysFromInPredicate(
       final List<InPredicate> inPredicates,
       final LogicalSchema schema,
       final MetaStore metaStore,
       final KsqlConfig config
   ) {
+    if (schema.key().size() > 1) {
+      throw invalidWhereClauseException("Schemas with multiple "
+          + "KEY columns are not supported", false);
+    }
+
     final InPredicate inPredicate = Iterables.getLast(inPredicates);
     final Column keyColumn = Iterables.getOnlyElement(schema.key());
     return inPredicate.getValueList()
         .getValues()
         .stream()
         .map(expression -> resolveKey(expression, keyColumn, metaStore, config, inPredicate))
+        .map(GenericKey::genericKey)
         .collect(Collectors.toList());
   }
 
-  private static Object extractKeyWhereClause(
+  private static GenericKey extractKeyWhereClause(
       final List<ComparisonExpression> comparisons,
       final boolean windowed,
       final LogicalSchema schema,
       final MetaStore metaStore,
       final KsqlConfig config
   ) {
-    final ComparisonExpression comparison = Iterables.getLast(comparisons);
-    if (comparison.getType() != Type.EQUAL) {
-      final ColumnName keyColumn = Iterables.getOnlyElement(schema.key()).name();
-      throw invalidWhereClauseException("Bound on '" + keyColumn.text()
-          + "' must currently be '='", windowed);
+    final Object[] keyContents = new Object[schema.key().size()];
+    final BitSet seenKeys = new BitSet(schema.key().size());
```

Comment:
I'm not super familiar with BitSet, but is this the number of unique bits or the highest bit value you have to account for?

And are key column indexes guaranteed to be the first 0...N indexes or can they skip around?

I'm just asking to to be sure this has the capacity and because I'm curious to learn the answers.  :-)

Reply:
```java
/**
     * Creates a bit set whose initial size is large enough to explicitly
     * represent bits with indices in the range {@code 0} through
     * {@code nbits-1}. All bits are initially {@code false}.
     *
     * @param  nbits the initial size of the bit set
     * @throws NegativeArraySizeException if the specified initial size
     *         is negative
     */
    public BitSet(int nbits) {
```
A bitset is essentially `boolean[]` but with one bit per boolean - I'm not entirely sure what the difference is between "number of unique bits or the highest bit value you have to account for".

> And are key column indexes guaranteed to be the first 0...N indexes or can they skip around?
When you take it from `schema.key()` they are indexed as a sepearate namespace from the values. So there is a 0-indexed key and a 0-indexed value.

Reply:
Ah, ok.  that makes sense.  I didn't realize that.  If they weren't compacted like that, the index could be greater than the number of keys (and hence my mentioning unique keys vs highest index), but that's not the case here.